src/ManageCourses.Api/Data/UserDataHelper.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using GovUk.Education.ManageCourses.Api.Mapping;
+using GovUk.Education.ManageCourses.Domain.DatabaseAccess;
+using GovUk.Education.ManageCourses.Domain.Models;
+
+namespace GovUk.Education.ManageCourses.Api.Data
+{
+    /// <summary>
+    /// This helper class generates the diff between the existing and new dataset.
+    /// Then implements the upsert
+    /// </summary>
+    public class UserDataHelper
+    {
+        private readonly IManageCoursesDbContext _context;
+        private readonly IReadOnlyCollection<McUser> _dataList;
+        private ImportMapper<McUser> _mapper;
+        /// <summary>
+        /// constructor for the class. Initialises the dataset and context.
+        /// Generates the diff (additions, updates, deletions)
+        /// </summary>
+        /// <param name="context">context for the db</param>
+        /// <param name="dataList">new dataset coming from the importer</param>
+        public UserDataHelper(IManageCoursesDbContext context, IReadOnlyCollection<McUser> dataList)
+        {
+            _context = context;
+            _dataList = dataList;
+            GenerateDiff();
+        }
+        /// <summary>
+        /// Actions the additions, deletions, updates
+        /// </summary>
+        /// <returns></returns>
+        public bool Upsert()
+        {
+            var returnBool = false;
+            try
+            {
+                _context.McUsers.AddRange(_mapper.Additions);
+                _context.McUsers.RemoveRange(_mapper.Deletes);
+                DoUpdates();
+                _context.Save();
+                returnBool = true;
+            }
+            catch (Exception e)
+            {
+                //write to error log?
```

Comment:
Transaction maybe?

Reply:
An UpsertResult class has been added that holds an error message or upsert details if successful